contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,846 @@
+ Vertical scaling must be performed destructively.
+ - The primary means of configuration update will be configuration files, 
+ command line flags, environment variables, or ConfigMaps consumed as the one 
+ of the former. 
+ - In place configuration update via SIGHUP is not universally 
+ supported, and Kubelet provides no mechanism to perform this currently. Pod 
+ reconfiguration will be performed destructively.
+ - Stateful applications are likely to evolve wire protocols and storage formats
+  between versions. In most cases, when updating the application's Pod's 
+  containers, it will not be safe to roll back or forward to an arbitrary 
+  version. Controller based Pod update should work well when rolling out an 
+  update, or performing a rollback, between two specific revisions of the 
+  controlled API object. This is how Deployment functions, and this property is,
+  perhaps, even more critical for stateful applications.
+
+## Requirements
+This design is based on the following requirements.
+- Users must be able to update the containers of a StatefulSet's Pods.
+  - Updates to container commands, images, resources and configuration must be 
+  supported.
+- The update must progress in a sequential, deterministic order and respect the 
+  StatefulSet
+  [identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+  [deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+  guarantees.
+- A failed update must halt.
+- Users must be able to rollback an update.
+- Users must be able to roll forward to fix a failing/failed update.
+- Users must be able to view the status of an update.
+- Users should be able to view a bounded history of the updates that have been 
+applied to the StatefulSet.
+
+## API Object
+
+The following modifications will be made to the StatefulSetSpec API object.
+
+```go
+type StatefulSetSpec struct {
+    // Replicas, Selector, Template, VolumeClaimsTemplate, and ServiceName 
+    // omitted for brevity.
+
+	// TemplateGeneration is a monotonically increasing, 64 bit, integer used to 
+	// indicate the version of the of the PodTemplateSpec. If nil, the 
+	// StatefulSetController has not initialized its revision history,
+	// change tracking is not enabled, and all Pods will be created from 
+	// Template.
+	TemplateGeneration  *int64 `json:"templateGeneration"`
+
+	// GenerationPartition partitions the Pods in the StatefulSet by ordinal such 
+    // that all Pods with a lower ordinal will be created from the PodTemplate that
+    // represents the current revision of the StatefulSet's revision history and 
+    // all Pods with an a greater or equal ordinal will be created from the 
+    // PodTemplate that represents the target revision of the StatefulSet's 
+    // revision history.
+	GenerationPartition *int32 `json:"generationPartition,omitempty`
```

Comment:
who manipulates this field ? Can i use this field to stay i want to rollout a change to StatefulSet for only 1 pod and then rollback if it doesnt work @kow3ns  ?